test/integration/shoots/applications/shoot_app_test.go

```diff
@@ -0,0 +1,332 @@
+	seedName          = flag.String("seedName", "", "the name of the seed for the test shoot")
+	shootName         = flag.String("shootName", "", "the name of the shoot we want to test")
+	shootNamespace    = flag.String("shootNamespace", "", "the namespace name that the shoot resides in")
+	testShootsPrefix  = flag.String("prefix", "", "prefix to use for test shoots")
+	logLevel          = flag.String("verbose", "", "verbosity level, when set, logging level will be DEBUG")
+	downloadPath      = flag.String("downloadPath", "/tmp/test", "the path to which you download the kubeconfig")
+	shootTestYamlPath = flag.String("shootpath", "", "the path to the shoot yaml that will be used for testing")
+	cleanup           = flag.Bool("cleanup", false, "deletes the newly created / existing test shoot after the test suite is done")
+)
+
+const (
+	GuestbookAppTimeout         = 10 * time.Minute
+	DownloadKubeconfigTimeout   = 10 * time.Minute
+	WaitForShootCreationTimeout = 3600 * time.Second
+	WaitForShootDeletionTimeout = 3600 * time.Second
+
+	APISERVER             = "kube-apiserver"
+	GuestBookTemplateName = "guestbook-app.yaml.tpl"
+)
+
+var _ = Describe("Shoot application testing", func() {
+	var (
+		shootGardenerTest   *framework.ShootGardenerTest
+		shootTestOperations *framework.GardenerTestOperation
+		tillerOptions       *framework.TillerOptions
+		shootAppTestLogger  *logrus.Logger
+		apiserverLabels     map[string]string
+		guestBooktpl        *template.Template
+		targetTestShoot     *v1beta1.Shoot
+		dashboardIP, nodeIP string
+	)
+
+	BeforeSuite(func() {
+		if len(*kubeconfig) == 0 {
+			Fail("The path to the Garden cluster kubeconfig is required, please specify it and try again")
+		}
+
+		shootAppTestLogger = logger.AddWriter(logger.NewLogger(*logLevel), GinkgoWriter)
+
+		tillerOptions = &framework.TillerOptions{
+			DeploymentName:    "tiller-deploy",
+			Namespace:         metav1.NamespaceSystem,
+			EnableHostNetwork: false,
+		}
+
+		// check if a shoot spec is provided, if yes create a shoot object from it and use it for testing
+		// alternatively, if the
+		if len(*shootTestYamlPath) > 0 {
+			// parse shoot yaml into shoot object and generate random test names for shoots
+			_, shootObject, err := framework.CreateShootTestArtifacts(*shootTestYamlPath, *testShootsPrefix)
+			Expect(err).NotTo(HaveOccurred())
+
+			shootGardenerTest, err = framework.NewShootGardenerTest(*kubeconfig, shootObject, shootAppTestLogger)
+			Expect(err).NotTo(HaveOccurred())
+
+			timeoutCtx, cancel := context.WithTimeout(context.Background(), WaitForShootCreationTimeout)
+			shootCreationCtx := context.WithValue(timeoutCtx, "name", "create shoot")
+			defer cancel()
+
+			targetTestShoot, err = shootGardenerTest.CreateTestShoot(shootCreationCtx)
+			Expect(err).NotTo(HaveOccurred())
+
+			shootTestOperations, err = framework.NewGardenTestOperationFromObject(shootGardenerTest, targetTestShoot, tillerOptions)
+			Expect(err).NotTo(HaveOccurred())
+		} else {
```

Comment:
Too many `if` / `else` - factor out blocks to functions and use early `return`s.

Reply:
What about this comment?

Reply:
This is still not resolved.